### PR TITLE
SILGen: simplify cleanups and avoid critical edges.

### PR DESCRIPTION
**Pull request by atrick**
Keep the CFG in a valid, critical-edge free state that the SIL passes can handle.
Replace the ad-hoc on-the-fly cleanup block optimization with a more general block merging.

This simplifies SILGen now and will in subsequent commits allow massive simplification of SIL APIs and SIL passes.

Note that there's no downside to avoiding critical edges since they will be split anyway any various points in the SIL pipeline. However, there are many passes and utilities that can be made much simpler and more compute time and memory efficient.

### CODE REVIEW

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -4436,7 +4438,6 @@ RValue RValueEmitter::visitIfExpr(IfExpr *E, SGFContext C) {
     
     // FIXME: We could avoid imploding and reexploding tuples here.
     Condition cond = SGF.emitCondition(E->getCondExpr(),
-                                       /*hasFalse*/ true,
```
Comment (gottesmm) [inline lib/SILGen/SILGenExpr.cpp]: I think this commit needs to go before the previous one, no? Beyond that looks great!

**Review by gottesmm — COMMENTED** [review]
Looks like you are dicing/splicing. I think I found some nice feedback. Want to give it to you before you slice and dice more.

File: lib/SILGen/SILGenFunction.h [hunk 2]
```diff
@@ -693,14 +691,15 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
...
+  /// If `contArgs` is nonempty, then both Condition::exitTrue() and
+  /// Condition::exitFalse() must be called.
+  Condition emitCondition(Expr *E, bool invertValue = false,
                           ArrayRef<SILType> contArgs = {},
                           ProfileCounter NumTrueTaken = ProfileCounter(),
                           ProfileCounter NumFalseTaken = ProfileCounter());
 
-  Condition emitCondition(SILValue V, SILLocation Loc, bool hasFalseCode = true,
-                          bool invertValue = false,
+  Condition emitCondition(SILValue V, SILLocation Loc, bool invertValue = false,
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILGen/SILGenFunction.h]: If that is always a true condition, could we put an assert into Condition's destructor to catch this?

**Review by gottesmm — COMMENTED** [review]
One last one.

File: lib/SILGen/Condition.h [hunk 3]
```diff
@@ -51,29 +50,34 @@ class LLVM_LIBRARY_VISIBILITY Condition {
             SILLocation L)
     : TrueBB(TrueBB), FalseBB(FalseBB), ContBB(ContBB), Loc(L)
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SILGen/Condition.h]: I added an assert.

File: lib/SILGen/Condition.cpp [hunk 4]
```diff
@@ -36,110 +31,22 @@ static SILLocation getContinuationLoc(SILBasicBlock &BB, SILLocation Fallback) {
...
 }
 
 SILBasicBlock *Condition::complete(SILGenFunction &SGF) {
-  // If there is no continuation block, it's because we
-  // constant-folded the branch.  The case-exit will have left us in a
-  // normal insertion state (i.e. not a post-terminator IP) with
-  // nothing to clean up after.
-  if (!ContBB) {
-    return SGF.B.getInsertionBB();
+  assert(!TrueBB && "enterTrue is always called.");
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SILGen/Condition.cpp]: Yeah, I added some comments.

**Comment by atrick** [comment]
@gottesmm thanks for looking at this. I'd like to add feedback from review as separate commits so you don't need to re-review anything, but in this case, you hadn't finished the first round, and rebasing was part of the feedback. So the commits are all new, but nothing has changed functionally. Unfortunately, github isn't listing the commits in the right order anymore--trust that they are in the right order on the branch. Incidentally, I think it's better to clone-for-review for anything non-trivial.

**Review by gottesmm — COMMENTED** [review]
I still think 66bc68151dcfcc9c45869f08fab4220054adf12e should be before the previous one. What you would do is do the same change but make it so that you can pass false in that place. Then all usages of Condition pass false and we can eliminate that unfortunate code... no? I think together the end state is fine though.

Also:

1. I wanted to give a +1 on a particular comment. It reads awesome! 
2. Found a typo.

Beyond that LGTM.

**Comment by atrick** [comment]
Rerunning benchmarks to see if the few changes were spurious. Locally, I see only improvements. In particular, CI reported a 10% regression on IterateData, but I see a 10% improvement (although it seems very flaky). The only reason for regression would be incidental change to inlining because the cost model doesn't handle unconditional branches sensibly.

For code size, there is a 0.1% decrease in the benchmark suite and  0.1% increase in the stdlib.

**Comment by atrick** [comment]
@swift-ci benchmark.

**Comment by atrick** [comment]
I see no measurable impact on compile time.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test.

**Pull request merged**